builder/src/lib.rs

```diff
@@ -0,0 +1,104 @@
+// Licensed under the Apache-2.0 license
+
+use std::fs;
+use std::io::{self, ErrorKind};
+use std::path::Path;
+use std::process::Command;
+
+use elf::endian::LittleEndian;
+
+fn other_err(e: impl Into<Box<dyn std::error::Error + Send + Sync>>) -> io::Error {
+    io::Error::new(ErrorKind::Other, e)
+}
+
+fn run_cmd(cmd: &mut Command) -> io::Result<()> {
+    let status = cmd.status()?;
+    if status.success() {
+        Ok(())
+    } else {
+        Err(io::Error::new(
+            ErrorKind::Other,
+            format!(
+                "Process {:?} {:?} exited with status code {:?}",
+                cmd.get_program(),
+                cmd.get_args(),
+                status.code()
+            ),
+        ))
+    }
+}
+
+pub fn build_firmware_elf(fw_crate_name: &str, bin_name: &str) -> io::Result<Vec<u8>> {
+    const WORKSPACE_DIR: &str = concat!(env!("CARGO_MANIFEST_DIR"), "/..");
+    const TARGET: &str = "riscv32imc-unknown-none-elf";
+    const PROFILE: &str = "firmware";
+
+    run_cmd(
+        Command::new(env!("CARGO"))
+            .current_dir(WORKSPACE_DIR)
+            .arg("build")
+            .arg("--target")
+            .arg(TARGET)
+            .arg("--features=emu,riscv")
+            .arg("--profile")
+            .arg(PROFILE)
+            .arg("-p")
+            .arg(fw_crate_name)
+            .arg("--bin")
+            .arg(bin_name),
+    )?;
+    fs::read(
+        Path::new(WORKSPACE_DIR)
+            .join("target")
+            .join(TARGET)
+            .join(PROFILE)
+            .join(bin_name),
+    )
+}
+
+pub fn build_firmware_rom(fw_crate_name: &str, bin_name: &str) -> io::Result<Vec<u8>> {
```

Comment:
Is this function tailored for  the test harness ROM  or any ROM?

Reply:
This function takes any crate that has sections targeting the ROM address space and converts into a raw binary of the ROM. You could use it with the "real ROM" or with one of the test harness ROMs.

We'll need a similar function for the FMC, which ultimately will involve some code-signing and end up generating the fw blob that goes into the mailbox.